include/deal.II/lac/solver_cg.h

```diff
@@ -645,7 +702,18 @@ namespace internal
         alpha                = r_dot_preconditioner_dot_r / p_dot_A_dot_p;
 
         x.add(alpha, p);
-        residual_norm = std::sqrt(std::abs(r.add_and_dot(-alpha, v, r)));
+
+        // compute the residual norm with preconditioned residual
+        if (use_default_residual)
+          {
+            residual_norm = std::sqrt(std::abs(r.add_and_dot(-alpha, v, r)));
+          }
+        // compute the residual norm with the un-preconditioned residual, i.e.
+        // compute l2 norm of Ax - b.
+        else
+          {
+            residual_norm = r.l2_norm();
+          }
```

Comment:
Can you explain what you're trying to achieve here? I don't understand this change yet. What is done in `r.add_and_dot(-alpha, v, r);` is the sequence of the following two steps:
```c++
  r.add(-alpha, v);
  residual_norm = r.l2_norm();
```
However, with your change you do just the `l2_norm()` but without doing the update of the residual, which I believe would yield wrong results. It might be that I'm missing something, but in general I would also think that the residual we choose in the default implementation is the unpreconditioned residual. One could use the preconditioned residual by using the factors that are used for beta and alpha, but I do not see that we would use that.

Reply:
Just to add on the previous message, the `add_and_dot` does not have to do with the preconditioned/unpreconditioned residual, but is merely a compute optimization (to reduce the memory access).

Reply:
Hi! You are completely right, the right code should be:
```
if (use_default_residual)
  {
    residual_norm = std::sqrt(std::abs(r.add_and_dot(-alpha, v, r)));
  }
else
  {
    auto tmp = r;
    A.vmult(tmp, x);
    tmp.add(-1., b);
    residual_norm = tmp.l2_norm();
  }
```
This is the difference between [explicit and implicit](https://en.wikipedia.org/wiki/Conjugate_gradient_method#Explicit_residual_calculation) residual calculation. The explicit one is less sensitive to round-off errors (however it can be slower to evaluate).

Reply:
Yes, now I see how this makes sense, thank you! With this I would agree to the overall direction of the PR. I can look more into it later.

Reply:
Great! I will then make the changes (also on the documentation). Thank you :)

Reply:
Hi @kronbichler ! I made the changes we discussed :)